### ISSUE TITLE
Goals printed out of order?

### ISSUE THREAD

**Issue by jonsterling** [freebroccolo/vscode-redprl#3]
It was hard to tell before because RedPRL itself was spitting out goals in an order that did not comport with their names; however, I have just changed RedPRL to use always names `0,1,...,n` when printing a proof state, so it is clearer what is going on.

In the VS Code window, I now see the following:
![screen shot 2016-10-31 at 8 09 11 pm](https://cloud.githubusercontent.com/assets/55057/19875788/2ad591e0-9fa6-11e6-8d40-c30482c72ddc.png)

Whereas the output from RedPRL itself is something like this:
![screen shot 2016-10-31 at 8 11 24 pm](https://cloud.githubusercontent.com/assets/55057/19875800/44b3a0de-9fa6-11e6-8cec-93d443f67de6.png)

**Comment by freebroccolo** [comment]
Yeah, I was noticing something weird with the ordering too. Over the last few commits I was messing with a goal stack to try and control that to some extent because I thought they were just being printed backward at first and I was trying to keep the `Remaining Obligations` entry at the top. But I just noticed that sometimes that doesn't get ordered as expected either.

I'll have to investigate some more. It might be a bug in that gnarly function somewhere or possibly a bug in the diagnostics display. Or maybe there is just no guarantee about how they actually get ordered in the display.

**Comment by freebroccolo** [comment]
Does it make a difference if you zero-pad the goals in redprl? In other words, `01` instead of `1`. Maybe there is something happening in the editor that is causing them to be sorted.

**Comment by jonsterling** [comment]
@freebroccolo Sorry for losing track of this—I tried padding the goal indices in RedPRL and it didn't help; the same strange sorting happened.

**Comment by jonsterling** [comment]
This comment (https://github.com/Microsoft/vscode/issues/1927#issuecomment-237367990) suggests that diagnostics are just sorted by range, which could explain the weird behavior here.

**Comment by freebroccolo** [comment]
Ah, that makes sense. There's probably not much we can do about it on the editor side in that case unless something is changed upstream.

What about adjusting the numbering given by RedPRL to account for this?

**Comment by jonsterling** [comment]
@freebroccolo I don't think changing the numbering produced by RedPRL would help (I've tried a number of variations and nothing has had any effect)—VS Code is sorting only by ranges, and nothing we put into the numbering printed by RedPRL would change that, unless I'm misunderstanding... What's more, even if VS Code sorted secondarily by diagnostic message, it would still not be what we want: as a string `3` is greater than `10`.

It seems like unless upstream they add the ability to specify the ordering of diagnostics (or, if they would fall back to the original order of the diagnostics array when ranges are the same), we probably cannot show goals as individual diagnostics.

What do you think?

**Comment by freebroccolo** [comment]
@jonsterling I'm not sure what to do about this. What I was thinking is that we could maybe adjust the number of the goals produced by RedPRL by also sorting (and possibly renumbering) by ranges. So instead of `[#10]` and `[#1]` you'd see `[#0], and `[#1]`.

There'd still be the message sorting issue you mentioned, but this is where padding the goal numbers should help I think, since `03` would be less than `10`.

If that doesn't work, we could go back to diagnostics with combined goals but we'd lose the nice stacked goal display for hovers. We might be able to get that back with a separate hover feature but if we did that we'd need to reduce the amount of information provided by the diagnostic. The hover widget shows a combination of whatever the hover feature (separate from diagnostics) provides, along with whatever diagnostics are in the range, and then some other stuff depending on context. So having the diagnostic with combined goals shown there would be too much in addition to the separate goals we'd have to produce.

Does that make sense?

**Comment by jonsterling** [comment]
I'm not sure I completely follow, but I'm about to board a plane. Part of the problem is, I think the names assigned to goals has no observable effect on ordering at all, but I might be misunderstanding your suggestion.

**Comment by freebroccolo** [comment]
I just did some experiments and I think I understand how it works now.

First, the diagnostics are ordered by range as you suggest. After range, the order is determined by the order in which the diagnostic is added to the diagnostics array.

Example:

```typescript
    this.diagnostics.set(
      document.uri,
      [
        new vs.Diagnostic(new vs.Range(0, 0, 1, 0), "[01]"),
        new vs.Diagnostic(new vs.Range(3, 0, 4, 0), "[10]"),
        new vs.Diagnostic(new vs.Range(3, 0, 4, 0), "[03]"),
        new vs.Diagnostic(new vs.Range(3, 0, 4, 0), "[04]"),
        new vs.Diagnostic(new vs.Range(0, 0, 1, 0), "[02]"),
      ],
    );
```

will display as:

<img width="160" alt="order" src="https://cloud.githubusercontent.com/assets/151197/20641106/475532fa-b3ae-11e6-92a5-120b1c32ff96.png">

**Comment by freebroccolo** [comment]
Okay, this is pretty weird. Looks to me like it might actually be a bug in the editor. Notice that for the first and second range, `[#0]` gets printed last. But for the 3rd range, it is displayed correctly.

diagnostics stack:
```json
[
  [
    "Remaining Obligations",
    "[#0]\n⊢ loop[x] : S1;\n",
    "[#1]\n⊢ loop[dim0] = base : S1;\n",
    "[#2]\n⊢ loop[dim1] = base : S1;\n"
  ],
  [
    "Remaining Obligations",
    "[#0]\nf : (x : bool) -> bool true;\n⊢ tt : bool;\n",
    "[#1]\nf : (x : bool) -> bool true;\nx : bool true;\n_ : x = f tt : bool;\n⊢ bool true;\n",
    "[#2]\n⊢ (x : bool) -> bool type;\n"
  ],
  [
    "Remaining Obligations",
    "[#0]\ns : S1 true;\n⊢ base : S1;\n",
    "[#1]\ns : S1 true;\n⊢ loop[x] : S1;\n",
    "[#2]\ns : S1 true;\n⊢ loop[dim0] = base : S1;\n",
    "[#3]\ns : S1 true;\n⊢ loop[dim1] = base : S1;\n",
    "[#4]\n⊢ S1 type;\n"
  ]
]
```

diagnostics display:

<img width="531" alt="order" src="https://cloud.githubusercontent.com/assets/151197/20641450/35018282-b3b5-11e6-8a9e-5d69dce1f470.png">

**Comment by freebroccolo** [comment]
I did some more experimenting and there is definitely some weird reordering going on by the editor that doesn't seem to be related to the order of the diagnostic in the stack or to content of the message.

I haven't tried debugging it but the actual diagnostics are set by the extension host [here](https://github.com/Microsoft/vscode/blob/8e3388f064743187743dd70cf8c324ce891c3b13/src/vs/workbench/api/node/extHostDiagnostics.ts#L45-L143).

**Comment by jonsterling** [comment]
@freebroccolo I was looking at that `extHostDiagnostics.ts` code too; frankly, it's gnarly enough that it would not surprise me is something strange was happening there... But it's hard to tell.

**Comment by freebroccolo** [comment]
Yeah, I'm just going to open an issue in the vscode repo and see what they have to say about it.

**Comment by jonsterling** [comment]
@freebroccolo thanks for opening that ticket!

I was thinking also that we may want to explore displaying the goals in a pane in a fully expanded form... The diagnostics view is really cool, but one disadvantage is that most *individual* goals require multiple lines to be readable.

**Comment by freebroccolo** [comment]
Hmm, that's a good point. Maybe the diagnostics view is not really a good fit for this.

We could open a panel to the side like the markdown preview and display stuff there. One thing I dislike about that approach is that it takes more space but perhaps I could get used to it. I suppose we could still display the goals on hover anyway. In fact, it might be better that way since we could use syntax highlighting for the goals (we can't do this for the automatic hover diagnostics afaik).

What would be ideal is if we could display the goals inline in the editor panel using some sort of custom collapsable goal widget. I think we might eventually be able to do that when they implement that custom code peak stuff that I linked to awhile ago.

**Comment by jonsterling** [comment]
Yeah... I'm thinking something like an optional side panel or something that constantly displays goals, with syntax highlighting, would be really sweet; ideally we could also continue to display the hover diagnostics, if a user prefers to work without the persistent goal display (e.g. on a narrow screen). 

One thing that would be cool is if that feature where hovering on an identifier highlights all occurrences could be carried over into the goals display as well—it would help the user identify at a glance the dependencies between subgoals...

I love the idea of this collapsible goal widget inline too; long-term, I think that sounds like the ideal solution, and we should think about doing it as soon as it becomes possible.

**Comment by freebroccolo** [comment]
> One thing that would be cool is if that feature where hovering on an identifier highlights all occurrences could be carried over into the goals display as well—it would help the user identify at a glance the dependencies between subgoals...

If we just use a static web page that might be tricky. However, I think we can use arbitrary scripts in those virtual document panels. In that case, we can probably embed the monaco editor (the editor component that vscode uses) which provides a lot of the same functionality (highlight occurrences, hover, definition preview):

![redprl](https://cloud.githubusercontent.com/assets/151197/20642254/62c5644c-b3c7-11e6-8f7c-7b18aedcd4f6.png)

However, in order to support some of the rich editing features in the goal display, we might need to also produce some diagnostic info specifically for that panel too.

**Comment by jonsterling** [comment]
@freebroccolo interesting... My thinking is that we should start with the simplest thing that could possibly work.
